lib/src/git.rs

```diff
@@ -60,6 +60,43 @@ fn local_branch_name_to_ref_name(branch: &str) -> String {
     format!("refs/heads/{branch}")
 }
 
+/// Checks if `git_ref` points to a Git commit object, and returns its id.
+///
+/// If the ref points to the previously `known_target` (i.e. unchanged), this
+/// should be faster than `git_ref.peel_to_commit()`.
+fn resolve_git_ref_to_commit_id(
+    git_ref: &git2::Reference<'_>,
+    known_target: Option<&RefTarget>,
+) -> Option<CommitId> {
+    // Try fast path if we have a candidate id which is known to be a commit object.
+    if let Some(RefTarget::Normal(id)) = known_target {
+        if matches!(git_ref.target(), Some(oid) if oid.as_bytes() == id.as_bytes()) {
+            return Some(id.clone());
+        }
+        if matches!(git_ref.target_peel(), Some(oid) if oid.as_bytes() == id.as_bytes()) {
+            // A hard tag. No need to peel to a tag object.
+            return Some(id.clone());
+        }
+        // A tag (according to ref name.) Try to peel one more level. This is slightly
+        // faster than recurse into peel_to_commit(). If we recorded a tag oid, we
+        // could skip this at all.
+        if let Some(Ok(tag)) = git_ref.is_tag().then(|| git_ref.peel_to_tag()) {
+            if tag.target_id().as_bytes() == id.as_bytes() {
+                // No need to peel recursively to a commit object.
```

Comment:
Maybe "// An annotated tag pointing to the already known target commit" or similar to clarify why we don't need to peel?

Reply:
Done.